src/virtual_secondary/virtualsecondary.cc

```diff
@@ -70,26 +70,37 @@ VirtualSecondary::VirtualSecondary(Primary::VirtualSecondaryConfig sconfig_in)
 
 bool VirtualSecondary::storeFirmware(const std::string& target_name, const std::string& content) {
   if (fiu_fail((std::string("secondary_install_") + getSerial().ToString()).c_str()) != 0) {
+    // consider changing this approach of the fault injection, since the current approach impacts the non-test code flow
+    // here as well as it doesn't test the installation failure on secondary from an end-to-end perspective as it
+    // injects an error on the middle of the control flow that would have happened if an installation error had happened
+    // in case of the virtual or the ip-secondary or any other secondary, e.g. add a mock secondary that returns an
+    // error to sendFirmware/install request we might consider passing the installation description message from
+    // Secondary, not just bool and/or data::ResultCode::Numeric
     return false;
   }
+
+  // TODO: it does not make much sense to read, pass via a function parameter to Virtual secondary
+  // and store the file that has been already downloaded by Primary
+  // Primary should apply ECU (primary, virtual, secondary) specific verification, download and installation logic in
+  // the first place
```

Comment:
Not sure I agree with this.

Reply:
Why ? What's your argument(s) ?
 I think, it make sense to download a file straight to a desired location/place or copy it from the primary place to the desired location eliminating passing a file content as a string parameter of a function/method call.

Reply:
>  I think, it make sense to download a file straight to a desired location/place or copy it from the primary place to the desired location eliminating passing a file content as a string parameter of a function/method call.

I agree with that statement. We shouldn't really be passing strings around like this. It should either be a file handle or a read handle to the storage. Do we have a ticket for that already? We should if we don't.

> Primary should apply ECU (primary, virtual, secondary) specific verification, download and installation logic in the first place

This is what I don't agree with. Specifically, the Primary cannot apply installation logic for a Secondary. Or what do you mean by "installation logic", actually? Perhaps I've misunderstood.

Reply:
> Do we have a ticket for that already? We should if we don't.
No, we don't a have a ticket specifically for this issue, but it could be part of the epic on the secondary and package manager refactoring as it most likely will require the secondary interface change.

Reply:
Sure. Can you make that ticket?

We successfully changed the API call for `OpenStoredTarget` to use `StorageTargetRHandle`, which it seems like we might be able do for virtual Secondaries. It probably won't be that simple for "real" Secondaries, though.

Reply:
API/interface for real and virtual secondaries should be the same, effectively one of the questions here how to turn sendFirmware(const std::string& data) into something more flexible and robust so we can pass a file handler/name  in one case, TLS certs and treehub in another and actual image data in some other...